src/loadRc.js

```diff
@@ -31,73 +41,32 @@ module.exports = function loadRc(
     return null;
   }
 
-  function parseExtensionlessRcFile(content: ?string): ?cosmiconfig$Result {
-    if (!content) return null;
-    const pasedConfig = options.rcStrictJson
-      ? parseJson(content, filepath)
-      : yaml.safeLoad(content, { filename: filepath });
-    return {
-      config: pasedConfig,
-      filepath,
-    };
-  }
-
   function loadRcWithExtensions() {
-    let foundConfig = null;
-    return funcRunner(readRcFile('json'), [
-      (jsonContent: ?string) => {
-        // Since this is the first try, config cannot have been found, so don't
-        // check `if (foundConfig)`.
-        if (jsonContent) {
-          const successFilepath = `${filepath}.json`;
-          foundConfig = {
-            config: parseJson(jsonContent, successFilepath),
-            filepath: successFilepath,
-          };
-        } else {
-          return readRcFile('yaml');
-        }
-      },
-      (yamlContent: ?string) => {
-        if (foundConfig) {
-          return;
-        } else if (yamlContent) {
-          const successFilepath = `${filepath}.yaml`;
-          foundConfig = {
-            config: yaml.safeLoad(yamlContent, { filename: successFilepath }),
-            filepath: successFilepath,
-          };
-        } else {
-          return readRcFile('yml');
-        }
-      },
-      (ymlContent: ?string) => {
-        if (foundConfig) {
-          return;
-        } else if (ymlContent) {
-          const successFilepath = `${filepath}.yml`;
-          foundConfig = {
-            config: yaml.safeLoad(ymlContent, { filename: successFilepath }),
-            filepath: successFilepath,
-          };
-        } else {
-          return readRcFile('js');
-        }
-      },
-      (jsContent: ?string) => {
-        if (foundConfig) {
-          return;
-        } else if (jsContent) {
-          const successFilepath = `${filepath}.js`;
-          foundConfig = {
-            config: requireFromString(jsContent, successFilepath),
-            filepath: successFilepath,
-          };
-        } else {
-          return;
-        }
-      },
-      () => foundConfig,
+    function loadExtension(
+      extension: string,
+      parse: (content: string, filename: string) => Object
+    ) {
+      const fpath = `${filepath}.${extension}`;
+      const parseRcFile = createParseFile(fpath, parse, options.ignoreEmpty);
+
+      // Check the result from the previous `loadExtension` invocation. If result
+      // isn't null, just return that.
+      return result => {
+        if (result != null) return result;
+
+        // Try to load the rc file for the given extension.
+        return funcRunner(readRcFile(extension), [parseRcFile]);
```

Comment:
Lets pass the `fpath` we constructed here instead of the extension which is used to build that again in `readRcFile`.